### PR TITLE
Ross test branch updated export and builtin edge case handling / fixed multiple bugs.

### PR DESCRIPTION
**Pull request by K2mies**

### CODE REVIEW

File: out
Comment (zoni527) [inline out:1]: I'll delete this

File: src/builtins/minishell_builtin_cd.c [hunk 1]
```diff
@@ -72,27 +88,25 @@ void	builtin_cd(t_minishell *data, t_token *builtin_token)
...
-		path = ft_ma_strdup(data->arena, "");
-	if (!(path[0] == '-' && path[1] == '\0'))
+	if (path != NULL && !(path[0] == '-' && path[1] == '\0'))
 		ms_setenv(data, "OLDPWD", old_path);
 	free(old_path);
-	if (path[0] == '~')
+	if (builtin_token && builtin_token->next && builtin_token->next->next)
+		ft_putendl_fd("minishell: cd: too many arguments", 2);
+	else if (ft_strncmp(data->raw_input, "cd \"\"", 5) == 0
+		&& (data->raw_input[6] == ' ' || data->raw_input[6] == '\0'))
```
Comment (zoni527) [inline src/builtins/minishell_builtin_cd.c:102]: Space is not the only metacharacter, there could be also any operator, a newline or a tab to split up into other tokens, so this won't catch those cases

Comment (zoni527) [reply src/builtins/minishell_builtin_cd.c:102]: ft_strchr(METACHARACTERS, raw_input[6]) might

File: src/builtins/minishell_builtin_dir.c [hunk 2]
```diff
@@ -57,3 +57,23 @@ int	change_dir(t_minishell *data, char *str)
...
+ * @param data	pointer to the main data struct
+ */
+char	*safe_getcwd(t_minishell *data)
+{
+	char	*cwd;
+
+	cwd = getcwd(NULL, 0);
+	if (cwd == NULL)
+	{
+		cwd = ft_calloc(1, sizeof(char));
```
Comment (zoni527) [inline src/builtins/minishell_builtin_dir.c:74]: Is this to prevent NULL dereferencing? Freeing NULL is completely safe, was this so that it can be safely used with strjoin or some such function?

File: src/minishell_main.c [hunk 3]
```diff
@@ -102,7 +103,7 @@ static const char	*get_prompt(t_minishell *data)
 	const char	*prompt;
 	char		*current_dir;
 
-	current_dir = getcwd(NULL, 0);
+	current_dir = safe_getcwd(data);
 	if (!current_dir)
```
Comment (zoni527) [inline src/minishell_main.c:107]: This NULL check is in conflict with safe_getcwd, it is explicitly programmed to not return NULL, so this check doesn't do anything.

**Review by zoni527 — COMMENTED** [review]
Need to check a few spots

File: src/minishell_main.c (hunk 3 above)
Comment (K2mies) [reply src/minishell_main.c:107]: get_prompt is your function, i just updated it to use safe_getcwd for posterity but didn't want to tamper with it otherwise. You can remove the double check.

File: src/builtins/minishell_builtin_dir.c (hunk 2 above)
Comment (K2mies) [reply src/builtins/minishell_builtin_dir.c:74]: its to cath the case in which getcwd returns NULL because otherwise cwd would be set explicitly to NULL and cause problems in the rest of the program. This was something we actually implemented together, i dunno if maybe you forgot about it?

File: src/minishell_main.c (hunk 3 above)
Comment (zoni527) [reply src/minishell_main.c:107]: I'll remove the check!

File: src/builtins/minishell_builtin_dir.c (hunk 2 above)
Comment (zoni527) [reply src/builtins/minishell_builtin_dir.c:74]: Oh yeah, I keep forgetting which functions are used where, so I didn't remember that this function's result is used to assign a t_vars value.

**Review by zoni527 — APPROVED** [review]
